lib/class-wp-rest-plugins-controller.php

```diff
@@ -0,0 +1,822 @@
+			);
+		}
+
+		$can_read = $this->check_read_permission( $request['plugin'] );
+
+		if ( is_wp_error( $can_read ) ) {
+			return $can_read;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Retrieves one plugin from the site.
+	 *
+	 * @since 5.5.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_item( $request ) {
+		require_once ABSPATH . 'wp-admin/includes/plugin.php';
+
+		$data = $this->get_plugin_data( $request['plugin'] );
+
+		if ( is_wp_error( $data ) ) {
+			return $data;
+		}
+
+		return $this->prepare_item_for_response( $data, $request );
+	}
+
+	/**
+	 * Checks if the given plugin can be viewed by the current user.
+	 *
+	 * On multisite, this hides non-active network only plugins if the user does not have permission
+	 * to manage network plugins.
+	 *
+	 * @since 5.5.0
+	 *
+	 * @param string $plugin The plugin file to check.
+	 * @return true|WP_Error True if can read, a WP_Error instance otherwise.
+	 */
+	protected function check_read_permission( $plugin ) {
+		if ( ! $this->is_plugin_installed( $plugin ) ) {
+			return new WP_Error( 'rest_plugin_not_found', __( 'Plugin not found.', 'gutenberg' ), array( 'status' => 404 ) );
+		}
+
+		if ( ! is_multisite() ) {
+			return true;
+		}
+
+		if ( ! is_network_only_plugin( $plugin ) || is_plugin_active( $plugin ) || current_user_can( 'manage_network_plugins' ) ) {
+			return true;
+		}
+
+		return new WP_Error(
+			'rest_cannot_view_plugin',
+			__( 'Sorry, you are not allowed to manage this plugin.', 'gutenberg' ),
+			array( 'status' => rest_authorization_required_code() )
+		);
+	}
+
+	/**
+	 * Checks if a given request has access to upload plugins.
+	 *
+	 * @since 5.5.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has access to create items, WP_Error object otherwise.
+	 */
+	public function create_item_permissions_check( $request ) {
+		if ( ! current_user_can( 'install_plugins' ) ) {
+			return new WP_Error(
+				'rest_cannot_install_plugin',
+				__( 'Sorry, you are not allowed to install plugins on this site.', 'gutenberg' ),
+				array( 'status' => rest_authorization_required_code() )
+			);
+		}
+
+		if ( 'inactive' !== $request['status'] && ! current_user_can( 'activate_plugins' ) ) {
+			return new WP_Error(
+				'rest_cannot_activate_plugin',
+				__( 'Sorry, you are not allowed to activate plugins.', 'gutenberg' ),
+				array(
+					'status' => rest_authorization_required_code(),
+				)
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Uploads a plugin and optionally activates it.
+	 *
+	 * @since 5.5.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function create_item( $request ) {
```

Comment:
Should we support uploading the plugin as file?

Reply:
I think that is something to consider in a future release.